src/platform/Zephyr/DiagnosticDataProviderImpl.cpp

```diff
@@ -29,17 +29,76 @@
 
 #include <drivers/hwinfo.h>
 
+#ifdef CONFIG_MCUBOOT_IMG_MANAGER
+#include <dfu/mcuboot.h>
+#endif
+
 #include <malloc.h>
 
 namespace chip {
 namespace DeviceLayer {
 
+namespace {
+
+uint8_t DetermineBootReason()
+{
+#ifdef CONFIG_HWINFO
+    uint32_t reason;
+
+    if (hwinfo_get_reset_cause(&reason) != 0)
+    {
+        return BootReasonType::Unspecified;
+    }
+
+    // Bits returned by hwinfo_get_reset_cause() are accumulated between subsequent resets, so
+    // the reset cause must be cleared after reading in order to make sure it always contains
+    // information about the most recent boot only.
+    (void) hwinfo_clear_reset_cause();
+
+    // If no reset cause is provided, it indicates a power-on-reset.
+    if (reason == 0 || reason & (RESET_POR | RESET_PIN))
+    {
+        return BootReasonType::PowerOnReboot;
+    }
+
+    if (reason & RESET_WATCHDOG)
+    {
+        return BootReasonType::HardwareWatchdogReset;
+    }
+
+    if (reason & RESET_BROWNOUT)
+    {
+        return BootReasonType::BrownOutReset;
+    }
+
+    if (reason & RESET_SOFTWARE)
+    {
+#ifdef CONFIG_MCUBOOT_IMG_MANAGER
+        if (mcuboot_swap_type() == BOOT_SWAP_TYPE_REVERT)
```

Comment:
I think this is a good enough, though i believe if another reset happens before we send NotifyApply then we won't send it anymore (at least immediately).

Reply:
But if another reset happens then MCUBoot will rollback the update, so it's good that we don't send NotifyUpdateApplied, right? We only need to make sure that OTA update will repeat within the next 24h.